rest/rest-resources/src/main/java/org/apache/brooklyn/rest/transform/TypeTransformer.java

```diff
@@ -0,0 +1,194 @@
+        }
+        if (item.getIconUrl()!=null) {
+            result.setIconUrl(tidyIconLink(b, item, item.getIconUrl(), ub));
+        }
+        
+        if (detail) {
+            if (RegisteredTypes.isSubtypeOf(item, Entity.class)) {
+                embellishEntity(result, item, b);
+            } else if (RegisteredTypes.isSubtypeOf(item, EntityAdjunct.class) ||
+                    // when implied supertypes are used we won't need the code below
+                    RegisteredTypes.isSubtypeOf(item, Policy.class) || RegisteredTypes.isSubtypeOf(item, Enricher.class) || RegisteredTypes.isSubtypeOf(item, Feed.class)
+                    ) {
+                try {
+                    Set<AdjunctConfigSummary> config = Sets.newLinkedHashSet();
+                    
+                    AbstractBrooklynObjectSpec<?,?> spec = b.getTypeRegistry().createSpec(item, null, null);
+                    for (final SpecParameter<?> input : spec.getParameters()){
+                        config.add(EntityTransformer.adjunctConfigSummary(input));
+                    }
+                    
+                    result.getExtraFields().put("config", config);
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+                    log.trace("Unable to create spec for "+item+": "+e, e);
+                }
+                
+            } else if (RegisteredTypes.isSubtypeOf(item, Location.class)) {
+                // TODO include config on location specs?  (wasn't done previously so not needed, but good for completeness)
+                result.getExtraFields().put("config", Collections.emptyMap());
+            }
+        }
+        return result;
+    }
+
+    protected static <T extends TypeSummary> void embellishEntity(T result, RegisteredType item, BrooklynRestResourceUtils b) {
+        try {
+            Set<EntityConfigSummary> config = Sets.newLinkedHashSet();
+            Set<SensorSummary> sensors = Sets.newTreeSet(SummaryComparators.nameComparator());
+            Set<EffectorSummary> effectors = Sets.newTreeSet(SummaryComparators.nameComparator());
+      
+            EntitySpec<?> spec = b.getTypeRegistry().createSpec(item, null, EntitySpec.class);
+            EntityDynamicType typeMap = BrooklynTypes.getDefinedEntityType(spec.getType());
+            EntityType type = typeMap.getSnapshot();
+   
+            AtomicInteger paramPriorityCnt = new AtomicInteger();
+            for (SpecParameter<?> input: spec.getParameters())
+                config.add(EntityTransformer.entityConfigSummary(input, paramPriorityCnt));
+            for (Sensor<?> x: type.getSensors())
+                sensors.add(SensorTransformer.sensorSummaryForCatalog(x));
+            for (Effector<?> x: type.getEffectors())
+                effectors.add(EffectorTransformer.effectorSummaryForCatalog(x));
+            
+            result.getExtraFields().put("config", config);
```

Comment:
Where do you imagine we'll document the names "config", "sensors" and "effectors" (or do people have to discover it by making some live calls and figuring out what looks to be there consistently for everything of a given type)?

Where would we document the type of the values for something like an effector?

I think the looseness won't be an issue for Brooklyn developers (because we can easily navigate the code base to find the `EffectorTransformer`, when working on the CLI or UI).

However, I think it makes it feel more scary for a user of Brooklyn who is thinking of writing an integration with our REST api. They won't know what fields to expect. They can make a few calls and then guess at what is optional versus mandatory, and then code as defensively as possible in case their guess was wrong.

Reply:
for now there is no doc for extra info for specific types.  agree this is a gap but it is a lot more classes and maintenance to have all that detail completely type specific.  i'm quite sure run-time discoverability for extra fields will be sufficient for a while -- and it's not at all uncommon in pragmatic REST APIs i have used -- and when users start wanting more precision then i think it's worth doing the extra work (and preference for extra fields added in this class with docs to say they are only for selected types, rather than lots of subtypes).